lib/ASTGen/Sources/ASTGen/Decls.swift

```diff
@@ -1,162 +1,503 @@
     }
 
-    let type = visit(node.type).rawValue
+    return .decl(declAndDeclContext.decl)
+  }
+
+  func visit(_ node: DeinitializerDeclSyntax) -> ASTNode {
+    let declAndDeclContext = DestructorDecl_create(
+      self.ctx,
+      self.declContext,
+      self.bridgedSourceLoc(for: node.deinitKeyword)
+    )
+
+    if let body = node.body {
+      self.visitFunctionCodeBlock(body, declContextAndDecl: declAndDeclContext)
+    }
 
-    return .decl(ParamDecl_create(ctx, loc, loc, firstName, loc, secondName, type, declContext))
+    return .decl(declAndDeclContext.decl)
   }
 
-  public func visit(_ node: FunctionDeclSyntax) -> ASTNode {
-    let staticLoc = bridgedSourceLoc(for: node)
-    let funcLoc = bridgedSourceLoc(for: node.funcKeyword)
-    let nameLoc = bridgedSourceLoc(for: node.identifier)
-    let rParamLoc = bridgedSourceLoc(for: node.signature.input.leftParen)
-    let lParamLoc = bridgedSourceLoc(for: node.signature.input.rightParen)
+  private func visitFunctionCodeBlock(_ block: CodeBlockSyntax, declContextAndDecl: BridgedDeclContextAndDecl) {
+    let oldDeclContext = declContext
+    declContext = declContextAndDecl.declContext
+    defer { declContext = oldDeclContext }
 
-    var nameText = node.identifier.text
-    let name = nameText.withBridgedString { bridgedName in
-      return ASTContext_getIdentifier(ctx, bridgedName)
+    AbstractFunctionDecl_setBody(declContextAndDecl.decl, self.visit(block).rawValue)
+  }
+}
+
+// MARK: - OperatorDecl
+
+extension BridgedOperatorFixity {
+  fileprivate init(tokenKind: TokenKind) {
+    switch tokenKind {
+    case .keyword(.infix): self = .infix
+    case .keyword(.prefix): self = .prefix
+    case .keyword(.postfix): self = .postfix
+    default: fatalError("Unknown operator fixity")
     }
+  }
+}
 
-    let returnType: ASTNode?
-    if let output = node.signature.output {
-      returnType = visit(output.returnType)
-    } else {
-      returnType = nil
+extension ASTGenVisitor {
+  func visit(_ node: OperatorDeclSyntax) -> ASTNode {
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
+    let (precedenceGroupName, precedenceGroupLoc) = self.bridgedIdentifierAndSourceLoc(
+      for: node.operatorPrecedenceAndTypes?.precedenceGroup
+    )
+
+    return .decl(
+      OperatorDecl_create(
+        self.ctx,
+        self.declContext,
+        BridgedOperatorFixity(tokenKind: node.fixity.tokenKind),
+        self.bridgedSourceLoc(for: node.operatorKeyword),
+        name,
+        nameLoc,
+        self.bridgedSourceLoc(for: node.operatorPrecedenceAndTypes?.colon),
+        precedenceGroupName,
+        precedenceGroupLoc
+      )
+    )
+  }
+}
+
+extension BridgedAssociativity {
+  fileprivate init(tokenKind: TokenKind) {
+    switch tokenKind {
+    case .keyword(.none): self = .none
+    case .keyword(.left): self = .left
+    case .keyword(.right): self = .right
+    default: fatalError("Unknown precedence group associativity value")
```

Comment:
Should this produce a diagnostic rather than failing? The incoming syntax tree could be ill-formed.

Reply:
Do we not expect a `SwiftParserDiagnostics` pass to catch this sort of thing? I.e. why do we want to allow for translation of ill-formed syntax trees?

Reply:
ASTGen should handle any swift-syntax tree, whether from the parser or not. There's a few cases that currently have `!` that really should not as well. @ahoppen as I know he has strong opinions on this :)

Reply:
Oh I get it know, this is about ill-formed trees, not source syntax. So what we want is a generic error about the syntax tree being invalid, I suppose? We don’t seem to currently have a way to propagate out an error either.

Reply:
> ASTGen should handle any swift-syntax tree, whether from the parser or not. There's a few cases that currently have ! that really should not as well.

Does SwiftSyntax have a way to verify nodes? We could have an interceptor or verifier pre-pass rather than having to recover from every corner where the model is too permissive in the middle of translation.

Reply:
> is there any value in attempting to recover from an invalid syntax tree?

There absolutely is! For example if you write `var x Int`, the parser will recover and introduce the `:` as a missing token and we should continue type checking based on that assumption to emit any diagnostics type checking might produce.

> It’s almost as if this is starting to resemble the mess that the legacy parser is

Yes, it’s unfortunate but these checks need to live somewhere 😬 

> Is there any reason we do not want a separate validation pass preceding translation (e.g. `sourceFile.validate()`), using the validation mechanics that are built in to SwiftSyntax?

If we split validation and the translation into the C++ AST into two separate phases, the two phases will inevitably diverge and we’ll end up in situations where ASTGen makes assumptions about things that were should have been detected in `validate()`, which introduces some kind of statefulness to the phase of the compiler at which you are viewing the SwiftSyntax tree. And IMO that’s a major problem with the C++ tree at the moment: Type checking / SILGen / … all make assumptions about which shape the tree might be in at that point and these assumptions to get broken on a regular basis, usually resulting in assertion failures / crashes.

Reply:
> There absolutely is! For example if you write `var x Int`, the parser will recover and introduce the : as a missing token and we should continue type checking based on that assumption to emit any diagnostics type checking might produce.

I was thinking of non-parsed trees fed by a fuzzer or something, but I forgot you mentioned macro expansions as a possible input too. It makes perfect sense to recover when parsing written code though.

> If we split validation and the translation into the C++ AST into two separate phases, the two phases will inevitably diverge and we’ll end up in situations where ASTGen makes assumptions about things that were should have been detected in `validate()`.

My point is more about reusing existing validation logic. Even starting each `visit` method with a SwiftSyntax validation and error handling the result would be so much cleaner than reinventing it the way we do here. It feels sensible to me for ASTGen to trust SwiftSyntax and assume that it is being handed a valid structure per the language grammar. Personally, I think this assumption is less risky in the long run than the assumption that the SwiftSyntax and ASTGen validations are consistent.

Reply:
> My point is more about reusing existing validation logic. Even starting each `visit` method with a SwiftSyntax validation and error handling the result would be so much cleaner than reinventing it the way we do here. It feels sensible to me for ASTGen to trust SwiftSyntax and assume that it is being handed a valid structure per the language grammar. Personally, I think this assumption is less risky in the long run than the assumption that the SwiftSyntax and ASTGen validations are consistent.

Whenever we make assumptions on the syntax tree, I think those should be implemented in the type system and in many cases, I believe we succeed in that. Operator declaration parsing is a really weird one because of all the custom grammar that it has.

Reply:
> Whenever we make assumptions on the syntax tree, I think those should be implemented in the type system

Definitely!

> Operator declaration parsing is a really weird one because of all the custom grammar that it has

Do you mean precedence group parsing? Did you consider replacing `groupAttributes` with an optional child for every of the 4 kinds of attributes, and thus encoding that they must appear at most once?

Reply:
> Do you mean precedence group parsing? 

Oh, yes, sorry, I mixed them up.

> Did you consider replacing `groupAttributes` with an optional child for every of the 4 kinds of attributes, and thus encoding that they must appear at most once?

Not so far, but I think that would make sense.

Reply:
Ah, right. How ugly would it be to not actually enforce their order in the layout? Say, the initializer always inserts the attributes in a fixed order, but one could pass `nil`s in the initializer and then insert the attributes in the appropriate order via the setters if said order is important. The getters and setters would then work with a specific subrange of the layout of size 4 which is nilled out by default.

Reply:
I’m not entirely sure what you mean. Do you have a short example of how you think the API should look like?

Reply:
An optional child for each attribute kind as originally suggested, with accessors similar to these: 
```swift
  public var higherThanRelation: PrecedenceGroupRelationSyntax? {
      get {
        for index in 11...14 {
          guard let childData = data.child(at: index) else {
            // No data for this attribute (attributes are stored contiguously, i.e., no 'nil' data in between attributes).
            break
          }

          if let value = PrecedenceGroupRelationSyntax(Syntax(childData)) {
            return value
          }
        }

        return nil
      }

      set(value) {
        for index in 11...14 {
          guard let rawChild = data.rawChild(at: index) else {
            // No data for this attribute (attributes are stored contiguously, i.e., no 'nil' data in between attributes).
            // Stick it in at this index.
            self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: value?.data, arena: SyntaxArena()))
            return
          }

          if rawChild.kind == .precedenceGroupRelation {
            if let value {
              // Simply replace the data at this index.
              self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: value.data, arena: SyntaxArena()))
              return
            } else {
              self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: nil, andShiftingItTo: 14, arena: SyntaxArena()))
            }
          }
        }
      }
    }
  }
```

Another possibility is to statically feature the 4 optional child attributes in `PrecedenceGroupAttributeListSyntax`, as they would have been in `PredenceGroupDeclSyntax`, perhaps retaining the conformance to `SyntaxCollection`. This way no changes to `PredenceGroupDeclSyntax` are necessary.

Reply:
> An optional child for each attribute kind as originally suggested, with accessors similar to these.

The issue I see with this design is that a client of SwiftSyntax might create a `PrecedenceGroupDeclSyntax` that has two `higherThan` relations in the `groupAttributes` and the `higherThanRelation` getter would only return the first.

Similarly, if you set a `higherThan` relation using this setter, you can’t control where it is being inserted if no `higherThan` relation exists yet. So

```swift
let pg = PrecedenceGroupDeclSyntax()
pg.higherThanRelation = // …
pg.assignment = // ...
```

would give you a different node than

```swift
let pg = PrecedenceGroupDeclSyntax()
pg.assignment = // ...
pg.higherThanRelation = // …
```

and that’s quite surprising IMO.